### PR TITLE
Statistics SPI changes

### PR DESCRIPTION
**Pull request by losipiuk**
Change SPI for further statistics/cost work:
* change nulls count to nulls fraction
* introduce support for single value range to model low and high value in table.

This is one of many small PRs originating from original "statistics followup PR"

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java [hunk 1]
```diff
@@ -157,11 +157,18 @@ private Estimate calculateNullsCount(Map<String, PartitionStatistics> statistics
...
-    }
 
+        if (totalNullsCount.isValueUnknown() || totalRowsCount.isValueUnknown()) {
+            return Estimate.unknownValue();
+        }
+        if (totalRowsCount.getValue() == 0.0) {
+            return new Estimate(0.0);
+        }
+        return new Estimate(totalNullsCount.getValue() / totalRowsCount.getValue());
+    }
```
Comment (kokosing) [inline presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: put a blink line here

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/RangeColumnStatistics.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+import java.util.Optional;
+
+import static com.facebook.presto.spi.statistics.Estimate.unknownValue;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+
+public class RangeColumnStatistics
+{
+    public static final String DATA_SIZE_STATISTIC_KEY = "data_size";
+    public static final String FRACTION_STATISTICS_KEY = "fraction";
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/RangeColumnStatistics.java]: Which fraction? Values or nulls?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 3]
```diff
@@ -11,56 +11,52 @@
...
-    private static final String DISTINCT_VALUES_STATITIC_KEY = "distinct_values_count";
+    private static final List<RangeColumnStatistics> SINGLE_UNKNOWN_RANGE_STATISTICS = singletonList(RangeColumnStatistics.builder().build());
 
-    private ColumnStatistics(Estimate dataSize, Estimate nullsCount, Estimate distinctValuesCount)
-    {
-        requireNonNull(dataSize, "dataSize can not be null");
-        statistics = createStatisticsMap(dataSize, nullsCount, distinctValuesCount);
-    }
+    private final List<RangeColumnStatistics> rangeColumnStatistics;
+    private final Estimate nullsFraction;
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: Maybe it needs to remain a `Map`. Why was `Map` was used to keep statistics previously? Does it matter do you have tree or single entry?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/RangeColumnStatistics.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    public static final String DISTINCT_VALUES_STATISTICS_KEY = "distinct_values_count";
+
+    private final Optional<Object> lowValue;
+    private final Optional<Object> highValue;
+    private final Map<String, Estimate> statistics;
+
+    public RangeColumnStatistics(
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/RangeColumnStatistics.java]: I do not think it should be public. IMO it should only called from `Builder`.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/RangeColumnStatistics.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+public class RangeColumnStatistics
+{
+    public static final String DATA_SIZE_STATISTIC_KEY = "data_size";
+    public static final String FRACTION_STATISTICS_KEY = "fraction";
+    public static final String DISTINCT_VALUES_STATISTICS_KEY = "distinct_values_count";
+
+    private final Optional<Object> lowValue;
+    private final Optional<Object> highValue;
+    private final Map<String, Estimate> statistics;
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/RangeColumnStatistics.java]: What is a point of keeping these in map?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 6]
```diff
@@ -11,56 +11,52 @@
...
     {
-        Map<String, Estimate> statistics = new HashMap<>();
-        statistics.put(DATA_SIZE_STATISTIC_KEY, dataSize);
-        statistics.put(NULLS_FRACTION_STATISTIC_KEY, nullsFraction);
-        statistics.put(DISTINCT_VALUES_STATITIC_KEY, distinctValuesCount);
-        return unmodifiableMap(statistics);
-    }
+        this.nullsFraction = requireNonNull(nullsFraction, "nullsFraction can not be null");
+        requireNonNull(rangeColumnStatistics, "rangeColumnStatistics can not be null");
+        if (rangeColumnStatistics.size() > 1) {
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: Anyway we need to modify this class once we move to multiple ranges, so I do not see a point in storing them in list.

IMO when constructors are private, then `Builder` provides enough flexibility.

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java [hunk 7]
```diff
@@ -157,11 +157,18 @@ private Estimate calculateNullsCount(Map<String, PartitionStatistics> statistics
...
+                        return OptionalDouble.of(allPartitionsCount / partitionsWithStatisticsCount * nullsCount);
                     }
                 });
-    }
 
+        if (totalNullsCount.isValueUnknown() || totalRowsCount.isValueUnknown()) {
+            return Estimate.unknownValue();
+        }
+        if (totalRowsCount.getValue() == 0.0) {
+            return new Estimate(0.0);
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: `Estimate.zeroValue()` ?

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java [hunk 8]
```diff
@@ -183,13 +190,17 @@ private Estimate calculateNullsCountForPartitioningKey(HiveColumnHandle partitio
             return Estimate.unknownValue();
         }
 
-        return new Estimate(partitions.stream()
+        double estimatedTotalRowsCount = rowsPerPartition.getAsDouble() * partitions.size();
+        if (estimatedTotalRowsCount == 0.0) {
+            return Estimate.zeroValue();
+        }
+        double estimatedNullsCount = partitions.stream()
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: This is counted as long so perhaps make `estimatedNullsCount` a `long`, deferring cast to more visible place.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 9]
```diff
@@ -34,11 +34,11 @@ private ColumnStatistics(Estimate dataSize, Estimate nullsCount, Estimate distin
         statistics = createStatisticsMap(dataSize, nullsCount, distinctValuesCount);
     }
 
-    private static Map<String, Estimate> createStatisticsMap(Estimate dataSize, Estimate nullsCount, Estimate distinctValuesCount)
+    private static Map<String, Estimate> createStatisticsMap(Estimate dataSize, Estimate nullsFraction, Estimate distinctValuesCount)
     {
         Map<String, Estimate> statistics = new HashMap<>();
         statistics.put(DATA_SIZE_STATISTIC_KEY, dataSize);
-        statistics.put(NULLS_COUNT_STATISTIC_KEY, nullsCount);
+        statistics.put(NULLS_FRACTION_STATISTIC_KEY, nullsFraction);
```
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: i know no change here, can any of values being put be null? Maybe in such a case it's better not to put such a value at all? Otherwise annotate params with `@Nullable` (here and in `ColumnStatistics` ctor). Oh, in any case, annotate the getters as `@Nullable`.

(i see now you changed this not to be a map later on, but most of the comment still holds)

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java [hunk 10]
```diff
@@ -84,14 +85,19 @@ public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTab
...
+                rangeStatistics.setDistinctValuesCount(countDistinctPartitionKeys(hiveColumnHandle, hivePartitions));
+                nullsFraction = calculateNullsFractionForPartitioningKey(hiveColumnHandle, hivePartitions, partitionStatistics);
             }
             else {
-                columnStatistics.setDistinctValuesCount(calculateDistinctValuesCount(partitionStatistics, columnName));
-                columnStatistics.setNullsFraction(calculateNullsFraction(partitionStatistics, columnName, rowCount));
+                rangeStatistics.setDistinctValuesCount(calculateDistinctValuesCount(partitionStatistics, columnName));
+                nullsFraction = calculateNullsFraction(partitionStatistics, columnName, rowCount);
             }
+            columnStatistics.setNullsFraction(nullsFraction);
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: move `ColumnStatistics.Builder columnStatistics = ColumnStatistics.builder();` here

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java [hunk 11]
```diff
@@ -84,14 +85,19 @@ public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTab
...
+            rangeStatistics.setFraction(nullsFraction.map(value -> 1.0 - value));
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: move it just after if-else

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java [hunk 12]
```diff
@@ -195,16 +212,16 @@ private Node rewriteShowStats(ShowStats node, Table table, Constraint<ColumnHand
...
             TableStatistics tableStatistics = metadata.getTableStatistics(session, tableHandle, constraint);
             List<String> statisticsNames = findUniqueStatisticsNames(tableStatistics);
-            List<String> resultColumnNames = buildColumnsNames(statisticsNames);
-            List<SelectItem> selectItems = buildSelectItems(resultColumnNames);
-            Map<ColumnHandle, String> columnNames = getStatisticsColumnNames(tableStatistics, node, table.getName());
-
-            List<Expression> resultRows = buildStatisticsRows(tableStatistics, columnNames, statisticsNames);
+            List<String> statsColumnNames = buildStatsColumnsNames(statisticsNames);
+            List<SelectItem> selectItems = buildSelectItems(statsColumnNames);
+            Map<ColumnHandle, String> tableColumnNames = getStatisticsColumnNames(tableStatistics, tableHandle);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java]: renames `resultColumnNames -> statsColumnNames`, `columnNames -> tableColumnNames` could be a separate commit (but i don't insist)

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java [hunk 13]
```diff
@@ -239,15 +256,20 @@ protected Node visitNode(Node node, Void context)
...
             return statistics.getColumnStatistics()
                     .keySet().stream()
                     .collect(toMap(identity(), column -> metadata.getColumnMetadata(session, tableHandle, column).getName()));
         }
 
+        private Map<ColumnHandle, Type> getStatisticsColumnTypes(TableStatistics statistics, TableHandle tableHandle)
+        {
+            return statistics.getColumnStatistics()
+                    .keySet().stream()
+                    .collect(toMap(identity(), column -> metadata.getColumnMetadata(session, tableHandle, column).getType()));
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java]: Maybe you could squash `getStatisticsColumnNames` and `getStatisticsColumnTypes` (i don't understand their names btw) and return just `Map<ColumnHandle, ColumnMetadata>`?

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java [hunk 14]
```diff
@@ -260,23 +282,23 @@ private TableHandle getTableHandle(ShowStats node, QualifiedName table)
...
             statisticsKeys.addAll(tableStatistics.getTableStatistics().keySet());
             for (ColumnStatistics columnStats : tableStatistics.getColumnStatistics().values()) {
-                statisticsKeys.addAll(columnStats.getStatistics().keySet());
+                statisticsKeys.addAll(columnStats.getOnlyRangeColumnStatistics().getStatistics().keySet());
             }
             return unmodifiableList(new ArrayList(statisticsKeys));
         }
 
-        static List<Expression> buildStatisticsRows(TableStatistics tableStatistics, Map<ColumnHandle, String> columnNames, List<String> statisticsNames)
+        List<Expression> buildStatisticsRows(TableStatistics tableStatistics, Map<ColumnHandle, String> sourceColumnNames, Map<ColumnHandle, Type> sourceColumnTypes, List<String> statsColumnNames)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java]: `s/sourceColumnNames/tableColumnNames`, `s/sourceColumnTypes/tableColumnTypes`

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java [hunk 15]
```diff
@@ -286,33 +308,87 @@ private TableHandle getTableHandle(ShowStats node, QualifiedName table)
             return columnNames.stream().map(QueryUtil::unaliasedName).collect(toImmutableList());
         }
 
-        static List<String> buildColumnsNames(List<String> statisticsNames)
+        static List<String> buildStatsColumnsNames(List<String> statisticsNames)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java]: also make it private

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java [hunk 16]
```diff
@@ -286,33 +308,87 @@ private TableHandle getTableHandle(ShowStats node, QualifiedName table)
...
         {
             ImmutableList.Builder<String> columnNamesBuilder = ImmutableList.builder();
-            columnNamesBuilder.add("column_name");
-            columnNamesBuilder.addAll(statisticsNames);
+            columnNamesBuilder.add(COLUMN_NAME_COLUMN);
+            columnNamesBuilder.addAll(sortedCopyOf(
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java]: Is sorting important here? I know we want to be deterministic, but weren't we already?

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java [hunk 17]
```diff
@@ -286,33 +308,87 @@ private TableHandle getTableHandle(ShowStats node, QualifiedName table)
...
+                                    // we do not want to include "fraction" in show stats output if we do not have histograms
+                                    .filter(name -> !name.equals(FRACTION_STATISTICS_KEY))
+                                    .collect(toList()))
+                            .add(NULLS_FRACTION_COLUMN)
+                            .build()));
             return columnNamesBuilder.build();
         }
 
-        private static Row createStatsRow(Optional<String> columnName, List<String> statisticsNames, Map<String, Estimate> columnStatisticsValues)
+        private Row createColumnStatsRow(String columnName, Type columnType, ColumnStatistics columnStatistics, List<String> statsColumnNames)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java]: * could remain `static`
* `columnType` looks unused

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java [hunk 18]
```diff
@@ -286,33 +308,87 @@ private TableHandle getTableHandle(ShowStats node, QualifiedName table)
...
+                        break;
+                }
+            }
+            return new Row(rowValues.build());
+        }
 
-            rowValues.add(columnNameExpression);
-            for (String statName : statisticsNames) {
-                rowValues.add(createStatisticValueOrNull(columnStatisticsValues, statName));
+        private Expression asVarcharLiteral(Type valueType, Optional<Object> value)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java]: super cool, but unused

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java (hunk 3 above)
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: @kokosing the map survived as `com.facebook.presto.spi.statistics.RangeColumnStatistics#statistics`

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 19]
```diff
@@ -11,56 +11,52 @@
...
-        return statistics.get(DATA_SIZE_STATISTIC_KEY);
-    }
+        if (nullsFraction.isValueUnknown() ^ rangeColumnStatistics.get(0).getFraction().isValueUnknown()) {
+            throw new IllegalArgumentException("All or none fraction/nullsFraction must be set");
+        }
 
-    public Estimate getNullsFraction()
-    {
-        return statistics.get(NULLS_FRACTION_STATISTIC_KEY);
+        this.rangeColumnStatistics = unmodifiableList(rangeColumnStatistics);
```
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: Do defensive copy. Lack of Guava doesn't exempt you from being on the safe side :)

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 20]
```diff
@@ -11,56 +11,52 @@
...
+        }
+        if (rangeColumnStatistics.isEmpty()) {
+            rangeColumnStatistics = SINGLE_UNKNOWN_RANGE_STATISTICS;
+        }
 
-    public Estimate getDataSize()
-    {
-        return statistics.get(DATA_SIZE_STATISTIC_KEY);
-    }
+        if (nullsFraction.isValueUnknown() ^ rangeColumnStatistics.get(0).getFraction().isValueUnknown()) {
```
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: Nice, but `s/^/!=/`

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 21]
```diff
@@ -11,56 +11,52 @@
...
+    private ColumnStatistics(Estimate nullsFraction, List<RangeColumnStatistics> rangeColumnStatistics)
     {
-        Map<String, Estimate> statistics = new HashMap<>();
-        statistics.put(DATA_SIZE_STATISTIC_KEY, dataSize);
-        statistics.put(NULLS_FRACTION_STATISTIC_KEY, nullsFraction);
-        statistics.put(DISTINCT_VALUES_STATITIC_KEY, distinctValuesCount);
-        return unmodifiableMap(statistics);
-    }
+        this.nullsFraction = requireNonNull(nullsFraction, "nullsFraction can not be null");
+        requireNonNull(rangeColumnStatistics, "rangeColumnStatistics can not be null");
```
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: 👍 .. can the list contain null elements?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 22]
```diff
@@ -70,31 +66,42 @@ public static Builder builder()
...
+        public Builder addRange(Consumer<RangeColumnStatistics.Builder> rangeBuilderConsumer)
         {
-            this.nullsFraction = nullsFraction;
+            RangeColumnStatistics.Builder rangeBuilder = RangeColumnStatistics.builder();
+            rangeBuilderConsumer.accept(rangeBuilder);
+            addRange(rangeBuilder.build());
+            return this;
+        }
+
+        public Builder addRange(Object lowValue, Object highValue, Consumer<RangeColumnStatistics.Builder> rangeBuilderConsumer)
```
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: If you want to add a convenience overload for `addRange(Consumer<RangeColumnStatistics.Builder>`, consider having it without `Consumer<RangeColumnStatistics.Builder>`

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/RangeColumnStatistics.java [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.spi.statistics.Estimate.unknownValue;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+
+public class RangeColumnStatistics
```
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/RangeColumnStatistics.java]: `final` (in SPI this is not just a matter of boilerplate), esp. that the ctor is public

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/RangeColumnStatistics.java [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+            Optional<Object> highValue,
+            Estimate fraction,
+            Estimate dataSize,
+            Estimate distinctValuesCount)
+    {
+        this.lowValue = requireNonNull(lowValue, "lowValue can not be null");
+        this.highValue = requireNonNull(highValue, "highValue can not be null");
+        requireNonNull(fraction, "fraction can not be null");
+        requireNonNull(dataSize, "dataSize can not be null");
+        requireNonNull(distinctValuesCount, "distinctValuesCount can not be null");
```
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/RangeColumnStatistics.java]: you can defer the checks to "assignment" in `createStatisticsMap`

(unless you follow @kokosing advice (which i agree with) not to have a map here)

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/RangeColumnStatistics.java [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
+
+        public Builder setFraction(Estimate fraction)
+        {
+            this.fraction = fraction;
+            return this;
+        }
+
+        public Builder setDataSize(Estimate dataSize)
+        {
+            this.dataSize = requireNonNull(dataSize, "dataSize can not be null");
```
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/RangeColumnStatistics.java]: you don't validate in other builders method (and you don't have to)

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 26]
```diff
@@ -11,56 +11,52 @@
...
-    {
-        return statistics.get(NULLS_FRACTION_STATISTIC_KEY);
+        this.rangeColumnStatistics = unmodifiableList(rangeColumnStatistics);
     }
 
-    public Estimate getDistinctValuesCount()
+    public RangeColumnStatistics getOnlyRangeColumnStatistics()
     {
-        return statistics.get(DISTINCT_VALUES_STATITIC_KEY);
+        return rangeColumnStatistics.get(0);
```
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: i would put (repeat) the `size()==1` check here too, but probably it's against convention

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java (hunk 22 above)
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: it looks like not quite removed yet

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/RangeColumnStatistics.java [hunk 27]
```diff
@@ -0,0 +1,139 @@
...
+    {
+        this.lowValue = requireNonNull(lowValue, "lowValue can not be null");
+        this.highValue = requireNonNull(highValue, "highValue can not be null");
+        requireNonNull(fraction, "fraction can not be null");
+        requireNonNull(dataSize, "dataSize can not be null");
+        requireNonNull(distinctValuesCount, "distinctValuesCount can not be null");
+        this.statistics = createStatisticsMap(dataSize, fraction, distinctValuesCount);
+    }
+
+    private static Map<String, Estimate> createStatisticsMap(
```
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/RangeColumnStatistics.java]: what map?

**Comment by losipiuk** [comment]
addressed

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 28]
```diff
@@ -14,55 +14,35 @@
...
-        return statistics.get(NULLS_COUNT_STATISTIC_KEY);
+        return nullsCount;
     }
 
     public Estimate getDistinctValuesCount()
     {
-        return statistics.get(DISTINCT_VALUES_STATITIC_KEY);
-    }
-
-    public Map<String, Estimate> getStatistics()
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: It breaks backward-compatibility. Maybe we should leave for a while as deprecated?

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java [hunk 29]
```diff
@@ -130,9 +131,9 @@ private Estimate calculateDistinctValuesCount(Map<String, PartitionStatistics> s
                 DoubleStream::max);
     }
 
-    private Estimate calculateNullsCount(Map<String, PartitionStatistics> statisticsByPartitionName, String column)
+    private Estimate calculateNullsFraction(Map<String, PartitionStatistics> statisticsByPartitionName, String column, Estimate totalRowsCount)
     {
-        return summarizePartitionStatistics(
+        Estimate totalNullsCount = summarizePartitionStatistics(
```
Comment (kokosing) [inline presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: This function call is overly complicated to me, I would refactor that. Of course as a follow up.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java (hunk 6 above)
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: Both. I would make constructor which takes only a single value and then just store it as a field.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 30]
```diff
@@ -85,16 +84,6 @@ public Builder addRange(Consumer<RangeColumnStatistics.Builder> rangeBuilderCons
             return this;
         }
 
-        public Builder addRange(Object lowValue, Object highValue, Consumer<RangeColumnStatistics.Builder> rangeBuilderConsumer)
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: uff :)

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java [hunk 31]
```diff
@@ -255,65 +255,65 @@ private TableHandle getTableHandle(ShowStats node, QualifiedName table)
...
-                statisticsKeys.addAll(columnStats.getStatistics().keySet());
-            }
-            return unmodifiableList(new ArrayList(statisticsKeys));
+            ImmutableList.Builder<String> columnNamesBuilder = ImmutableList.builder();
+            columnNamesBuilder.add("column_name");
+            columnNamesBuilder.add("data_size");
+            columnNamesBuilder.add("distinct_values_count");
+            columnNamesBuilder.add("nulls_count");
+            columnNamesBuilder.add("row_count");
+            return columnNamesBuilder.build();
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java]: builder is better for so many key/values (i don't think there even is such an overload for `of`)

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 32]
```diff
@@ -76,7 +78,7 @@ public static Builder builder()
 
         public Builder setDataSize(Estimate dataSize)
         {
-            this.dataSize = requireNonNull(dataSize, "dataSize can not be null");
```
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: Builders usually don't validate

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java (hunk 28 above)
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: Do you think anyone cares?

Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: I will drop this. I do not believe anybody used it externally already.

**Comment by losipiuk** [comment]
> Both. I would make constructor which takes only a single value and then just store it as a field.

I decided to leave it as is. My motivation is to have all the checks in constructor. Otherwise I would have to split that between `Builder` and constructor.

**Comment by losipiuk** [comment]
@martint Only SPI changes left in this PR. Please let us know if it looks fine to you.

**Comment by losipiuk** [comment]
I will merge that after Travis passes. Changes where already discussed conceptually on Slack. Also I do not believe anyone depends on stats in SPI so far as there is no use of that in master branch yet.

**Comment by kokosing** [comment]
Hey, just wait a bit I have one comment!

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 33]
```diff
@@ -11,38 +11,54 @@
...
     {
-        return dataSize;
+        this.nullsFraction = requireNonNull(nullsFraction, "nullsFraction can not be null");
+        requireNonNull(rangeColumnStatistics, "rangeColumnStatistics can not be null");
+        if (!rangeColumnStatistics.stream().allMatch(Objects::nonNull)) {
+            throw new NullPointerException("elements of rangeColumnStatistics can not be null");
+        }
+        if (rangeColumnStatistics.size() > 1) {
+            // todo add support for multiple ranges.
+            throw new IllegalArgumentException("Statistics for multiple ranges are not supported");
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: What if we could allow user to create multiple ranges, but we could then just merge them into one for a time being until we start to support them?

Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: That is an option. But IMO we can add that later. This will not change signatures of constructor/builder. Just some calls which result in exception now, will start passing.
